feat: Pagination component (BH-856) and getting-started example

nickcernis opened the pull request:
- Adds a Next/WPGraphQL-based Pagination component to display next/previous links.
- Demonstrates its use in the getting-started `category.tsx` template.

<img width="1447" alt="Screenshot 2021-02-22 at 14 21 07" src="https://user-images.githubusercontent.com/647669/108714008-49ef8c00-7519-11eb-8ded-4f74c7a674d4.png">

## To test
1. In WordPress, make sure you have 21 or more posts in the uncategorized category. (For example, via [wp post generate](https://developer.wordpress.org/cli/commands/post/generate/).) This is so you can see posts across three pages.
2. `npm run bootstrap && npm run dev`
3. Visit http://localhost:3000/category/uncategorized.

You should see clickable next/previous page links.

## Notes
- Uses [cursor-based pagination as recommended by WPGraphQL](https://www.wpgraphql.com/2020/03/26/forward-and-backward-pagination-with-wpgraphql/). The downside is that sites migrating from WP will lose `/page/2` style pagination URLs, because WPGraphQL does not support/encourage this.
- I implemented this in a way that does not depend on client-side requests, so that users do not have to fill in `NEXT_PUBLIC_WORDPRESS_URL`, and so that paginated URLs such as `http://localhost:3000/category/uncategorized/after/YXJyYXljb25uZWN0aW9uOjU3MTY5` are more easily cacheable.

nickcernis committed:
feat: Add a Pagination component to display links

This depends on Next and WPGraphQL. It assumes the
cursor-based pagination strategy described at
https://www.wpgraphql.com/2020/03/26/forward-and-backward-pagination-with-wpgraphql/.

Custom queries must be handled separately by the
page template displaying the posts. See an example
in headless-framework/examples/getting-started/wp-templates/category.tsx.

nickcernis committed:
chore: Do not pass pagination info to WPGraphQL

When the path is
/category/uncategorized/after/abc123, we want to
only send /category/uncategorized to GraphQL for
the purpose of determining the correct template
for the current page.

So strip the `/after/abc123` portion, leaving
/category/uncategorized.

This assumes we want to paginate post archives
using URLs in the format:

/category/uncategorized/[before|after]/[abc123].

Where abc123 is the ID of the last or first post
on the current page.

This is different to the WP-native /page/1 format
but more closely matches the more efficient
cursor-based pagination strategy at
https://www.wpgraphql.com/2020/03/26/forward-and-backward-pagination-with-wpgraphql/

It also allows pages to be fetched on the server
(and cached) without using client-side requests
to fetch additional content. This means a user
does not have to fill in NEXT_PUBLIC_WORDPRESS_URL
to get paginated content.

nickcernis committed:
feat: Add pagination to getting-started template

To demonstrate a paginated archive. Will show
next and previous page links on category
archives such as category/uncategorized.

nickcernis committed:
style: Add basic styles for the pagination links

nickcernis committed:
Merge branch 'canary' into pagination

nickcernis commented on examples/getting-started/wp-templates/category.tsx:
```diff
@@ -1,42 +1,90 @@
-  const urlParts = router.asPath.split('/').filter(Boolean);
-  const category = urlParts[1];
-  const posts = usePosts({
+const POSTS_PER_PAGE = 10;
```
This could come from WP's reading settings via a custom query or a future `useReadingSettings()` hook, but I hard-coded it here for now to keep it more simple.

trevanhetzel commented:
Dang, that [cursor-based pagination article](https://www.wpgraphql.com/2020/03/26/forward-and-backward-pagination-with-wpgraphql/) is an excellent read!

trevanhetzel commented on examples/getting-started/wp-templates/category.tsx:
```diff
@@ -1,42 +1,90 @@
+ * Converts a string or array into category URL parts assuming the
+ * form /category/uncategorized/[before|after]/[abc123].
+ */
+const getCategoryUrlParts = (path: string | string[]): CategoryUrlParts => {
```
Is there anyway to bake `CategoryUrlParts`, `getCategoryUrlParts`, and `getQueryOptions` into the `Pagination` component itself? This is a lot of boilerplate. I would so love to just import `Pagination` from `@wpengine/headless` and literally use it with the framework just intrinsically knowing about the current cursor.

nickcernis replied:
Thanks for the feedback! I've refactored in https://github.com/wpengine/headless-framework/pull/154/commits/3b084b9bb8bec883510a6f2d6991562ecb014b46 to move the boilerplate out of the template. People don't need to know about pagination or cursors now.

trevanhetzel commented on packages/headless/src/next/Pagination.tsx:
```diff
@@ -0,0 +1,58 @@
+  if (pageInfo.hasNextPage) classes += ' has-next';
+  if (pageInfo.hasPreviousPage) classes += ' has-previous';
+
+  return (
```
As it stands right now, the component seems to mostly be a presentational component, with all the business logic living in the example file. This is fine, because people will likely want to use their own presentational component eventually, but I wonder if the business logic then could also be a component so as not to clutter the page template?

nickcernis replied:
Added a `categoryOptions` helper in https://github.com/wpengine/headless-framework/pull/154/commits/3b084b9bb8bec883510a6f2d6991562ecb014b46.

nickcernis committed:
refactor: extract category option logic

Puts category option logic in the framework
instead of in the page template.

As suggested by @trevanhetzel in
https://github.com/wpengine/headless-framework/pull/154.

nickcernis committed:
docs: improve getStaticProps info

nickcernis commented:
Refactored to move category option logic out of the template as suggested by @trevanhetzel.

This is ready for another look.

The ESLint CI failure seems does not seem related to this PR.

nickcernis committed:
chore: lint

nickcernis committed:
chore: lint

nickcernis committed:
refactor: 'keyword' to 'direction'

Keyword sounds like it relates to search terms.

Thanks to @mindctrl for the feedback.

mindctrl reviewed (approved):
LGTM. Works as expected. Per Slack chat, we ran into some issues that appear to be WPGraphQL issues.

https://github.com/wp-graphql/wp-graphql/issues/1552

https://github.com/wp-graphql/wp-graphql/issues/1714

Same behavior is visible in the [WPGraphQL example](https://codesandbox.io/s/wpgraphql-apollo-forward-backward-pagination-example-spq53?from-embed).

Given all that, I think we can merge this without any workarounds and work on fixes upstream.
